### PR TITLE
Service Mesh support

### PR DESCRIPTION
**Pull request by baluchicken**
| Q               | A
| --------------- | ---
| Bug fix?        |no|
| New feature?    |yes|
| API breaks?     |no|
| Deprecations?   |no|
| License         | Apache 2.0

### What's in this PR?
This PR adds support for
- Using unique service/broker to work with Istio
- Ability to disable external access to the Kafka Cluster
- Ability to use your own Cruise Control
- Use the newly released jmx exporter 0.12
- Remove all hard coded docker images

### Checklist

- [x] Implementation tested
- [x] Error handling code meets the [guideline](https://github.com/banzaicloud/pipeline/blob/master/docs/error-handling-guide.md)
- [x] Logging code meets the guideline
- [x] User guide and development docs updated (if needed)

### CODE REVIEW

File: pkg/apis/banzaicloud/v1alpha1/kafkacluster_types.go [hunk 1]
```diff
@@ -39,17 +39,20 @@ type KafkaClusterSpec struct {
 
 // KafkaClusterStatus defines the observed state of KafkaCluster
 type KafkaClusterStatus struct {
-	BrokersState map[int32]*BrokerState `json:"brokersState,omitempty"`
+	BrokersState       map[int32]*BrokerState    `json:"brokersState,omitempty"`
+	CruiseControlState *CruiseControlTopicStatus `json:"cruiseControlState,omitempty"`
```
Comment (pbalogh-sa) [inline pkg/apis/banzaicloud/v1alpha1/kafkacluster_types.go]: The type is confusing.

File: pkg/resources/cruisecontrol/deployment.go [hunk 2]
```diff
@@ -62,15 +61,10 @@ func (r *Reconciler) deployment(log logr.Logger) runtime.Object {
...
-						{
-							Name:    "create-topic",
-							Image:   "wurstmeister/kafka:2.12-2.1.0",
-							Command: []string{"/bin/bash", "-c", fmt.Sprintf("until /opt/kafka/bin/kafka-topics.sh --zookeeper %s --create --if-not-exists --topic __CruiseControlMetrics --partitions 12 --replication-factor 3; do echo waiting for kafka; sleep 3; done ;", strings.Join(r.KafkaCluster.Spec.ZKAddresses, ","))},
-						},
 						{
 							Name:    "jmx-exporter",
-							Image:   "banzaicloud/jmx_exporter:latest",
-							Command: []string{"cp", "/usr/share/jmx_exporter/jmx_prometheus_javaagent-0.3.1-SNAPSHOT.jar", "/opt/jmx-exporter/"},
+							Image:   "banzaicloud/jmx-javaagent:0.12.0",
```
Comment (pbalogh-sa) [inline pkg/resources/cruisecontrol/deployment.go]: The image should be configurable.

File: pkg/resources/cruisecontrol/deployment.go [hunk 3]
```diff
@@ -62,15 +61,10 @@ func (r *Reconciler) deployment(log logr.Logger) runtime.Object {
...
+							Command: []string{"cp", "/opt/jmx_exporter/jmx_prometheus_javaagent-0.12.0.jar", "/opt/jmx-exporter/"},
```
Comment (pbalogh-sa) [inline pkg/resources/cruisecontrol/deployment.go]: The version of javaagent should be configurable.

File: pkg/resources/cruisecontrol/deployment.go [hunk 4]
```diff
@@ -85,7 +79,7 @@ func (r *Reconciler) deployment(log logr.Logger) runtime.Object {
 							Env: []corev1.EnvVar{
 								{
 									Name:  "KAFKA_OPTS",
-									Value: "-javaagent:/opt/jmx-exporter/jmx_prometheus_javaagent-0.3.1-SNAPSHOT.jar=9020:/etc/jmx-exporter/config.yaml",
+									Value: "-javaagent:/opt/jmx-exporter/jmx_prometheus_javaagent-0.12.0.jar=9020:/etc/jmx-exporter/config.yaml",
```
Comment (pbalogh-sa) [inline pkg/resources/cruisecontrol/deployment.go]: The version of javaagent should be configurable.

File: pkg/resources/kafka/pod.go [hunk 5]
```diff
@@ -84,8 +82,8 @@ func (r *Reconciler) pod(broker banzaicloudv1alpha1.BrokerConfig, pvcs []corev1.
 				},
 				{
 					Name:    "jmx-exporter",
-					Image:   "banzaicloud/jmx_exporter:latest",
-					Command: []string{"cp", "/usr/share/jmx_exporter/jmx_prometheus_javaagent-0.3.1-SNAPSHOT.jar", "/opt/jmx-exporter/"},
+					Image:   "banzaicloud/jmx-javaagent:0.12.0",
```
Comment (pbalogh-sa) [inline pkg/resources/kafka/pod.go]: The image should be configurable.

File: pkg/resources/kafka/pod.go [hunk 6]
```diff
@@ -84,8 +82,8 @@ func (r *Reconciler) pod(broker banzaicloudv1alpha1.BrokerConfig, pvcs []corev1.
...
+					Command: []string{"cp", "/opt/jmx_exporter/jmx_prometheus_javaagent-0.12.0.jar", "/opt/jmx-exporter/"},
```
Comment (pbalogh-sa) [inline pkg/resources/kafka/pod.go]: The version of javaagent should be configurable.

File: pkg/resources/kafka/pod.go [hunk 7]
```diff
@@ -115,11 +113,15 @@ func (r *Reconciler) pod(broker banzaicloudv1alpha1.BrokerConfig, pvcs []corev1.
 						},
 						{
 							Name:  "KAFKA_OPTS",
-							Value: "-javaagent:/opt/jmx-exporter/jmx_prometheus_javaagent-0.3.1-SNAPSHOT.jar=9020:/etc/jmx-exporter/config.yaml",
+							Value: "-javaagent:/opt/jmx-exporter/jmx_prometheus_javaagent-0.12.0.jar=9020:/etc/jmx-exporter/config.yaml",
```
Comment (pbalogh-sa) [inline pkg/resources/kafka/pod.go]: The version of javaagent should be configurable.

File: pkg/resources/cruisecontrol/configmap.go [hunk 8]
```diff
@@ -37,7 +37,7 @@ func (r *Reconciler) configMap(log logr.Logger) runtime.Object {
     bootstrap.servers=%s:%d
     # The zookeeper connect of the Kafka cluster
     zookeeper.connect=%s/
-`, fmt.Sprintf(kafka.HeadlessServiceTemplate, r.KafkaCluster.Name), r.KafkaCluster.Spec.ListenersConfig.InternalListeners[0].ContainerPort, strings.Join(r.KafkaCluster.Spec.ZKAddresses, ",")) +
+`, fmt.Sprintf(kafka.AllBrokerServiceTemplate, r.KafkaCluster.Name), r.KafkaCluster.Spec.ListenersConfig.InternalListeners[0].ContainerPort, strings.Join(r.KafkaCluster.Spec.ZKAddresses, ",")) +
```
Comment (pepov) [inline pkg/resources/cruisecontrol/configmap.go]: what is the motivation behind this?

File: pkg/resources/cruisecontrol/cruisecontrol.go [hunk 9]
```diff
@@ -61,15 +61,30 @@ func (r *Reconciler) Reconcile(log logr.Logger) error {
...
-	for _, res := range []resources.ResourceWithLogs{
-		r.service,
-		r.configMap,
-		r.deployment,
-	} {
-		o := res(log)
-		err := k8sutil.Reconcile(log, r.Client, o, r.KafkaCluster)
+	if r.KafkaCluster.Spec.CruiseControlConfig.CruiseControlEndpoint == "" {
+
+		err := generateCCTopic(r.KafkaCluster)
```
Comment (pepov) [inline pkg/resources/cruisecontrol/cruisecontrol.go]: shouldn't we avoid calling this again once the status is `CruiseControlTopicReady`?

File: pkg/resources/cruisecontrol/configmap.go (hunk 8 above)
Comment (baluchicken) [reply pkg/resources/cruisecontrol/configmap.go]: Headless Service cause issues in Istio, especially in Multi Cluster environment, but we need a service which points to all of the brokers for Cruise Control.

File: pkg/resources/cruisecontrol/cruisecontrol.go (hunk 9 above)
Comment (baluchicken) [reply pkg/resources/cruisecontrol/cruisecontrol.go]: Yes you are right will do.

**Pull request merged**